File: shared/src/commonMain/kotlin/energy/ftw/protocol/Frame.kt
package energy.ftw.protocol

const val FRAME_VERSION = 1
const val HEADER_BYTES = 6
const val MAX_PAYLOAD = 0xffff
const val LANE_CONTROL = 0
const val LANE_BULK = 1
const val FLAG_TRUNC = 0x02
const val CONTROL_BUCKET = 512

data class Envelope(val t: String, val id: Long? = null, val b: Cbor? = null)

data class Frame(val lane: Int, val flags: Int, val envelope: Envelope)

class FrameError(message: String, val code: String) : Exception(message)

fun encodeFrame(frame: Frame, bucket: Int = CONTROL_BUCKET): ByteArray {
    val payload = encodeCbor(
        Cbor.map(
            *buildList {
                add("t" to Cbor.txt(frame.envelope.t))
                frame.envelope.id?.let { add("id" to Cbor.num(it)) }
                frame.envelope.b?.let { add("b" to it) }
            }.toTypedArray(),
        ),
    )
    if (payload.size > MAX_PAYLOAD) {
        throw FrameError("payload ${payload.size} exceeds u16 length field", "E_FRAME_TOO_LARGE")
    }
    val needed = HEADER_BYTES + payload.size
    if (needed > bucket) {
        throw FrameError("frame needs $needed bytes, bucket is $bucket", "E_FRAME_EXCEEDS_BUCKET")
    }
    val out = ByteArray(bucket)
    out[0] = FRAME_VERSION.toByte()
    out[1] = frame.lane.toByte()
    out[2] = frame.flags.toByte()
    out[3] = 0
    out[4] = (payload.size ushr 8).toByte()
    out[5] = payload.size.toByte()
    payload.copyInto(out, HEADER_BYTES)
    return out
}

fun decodeFrame(bytes: ByteArray): Frame {
    if (bytes.size < HEADER_BYTES) {
        throw FrameError("frame is ${bytes.size} bytes, need at least $HEADER_BYTES", "E_FRAME_SHORT")
    }
    val ver = bytes[0].toInt() and 0xff
    if (ver != FRAME_VERSION) {
        throw FrameError("unsupported frame version $ver", "E_FRAME_VERSION")
    }
    val lane = bytes[1].toInt() and 0xff
    val flags = bytes[2].toInt() and 0xff
    val len = ((bytes[4].toInt() and 0xff) shl 8) or (bytes[5].toInt() and 0xff)
    if (HEADER_BYTES + len > bytes.size) {
        throw FrameError("declared length $len overruns the ${bytes.size}-byte frame", "E_FRAME_TRUNCATED")
    }
    val payload = bytes.copyOfRange(HEADER_BYTES, HEADER_BYTES + len)
    val decoded = try {
        decodeCbor(payload)
    } catch (e: CborError) {
        throw FrameError("payload is not valid CBOR: ${e.message}", "E_FRAME_CBOR")
    }
    val map = decoded as? Cbor.Map ?: throw FrameError("envelope must be a CBOR map", "E_FRAME_ENVELOPE")
    val t = map.text("t") ?: throw FrameError("envelope is missing a string type", "E_FRAME_ENVELOPE")
    val id = map.long("id")
    val b = map["b"]
    return Frame(lane, flags, Envelope(t, id, b))
}

fun isTruncated(frame: Frame): Boolean = (frame.flags and FLAG_TRUNC) != 0

File: appleApp/FTWKit/Sources/FTWKit/Protocol/Frame.swift
import Foundation

/// The frame codec.
///
/// Each Noise transport message carries exactly one frame:
///
///     offset  field    type     note
///     0       ver      u8       frame layout version
///     1       lane     u8       0 = telemetry/control, 1 = bulk
///     2       flags    u8       0x02 TRUNC
///     3       rsvd     u8       0
///     4       len      u16 BE   payload bytes
///     6       payload  u8[len]  CBOR
///     6+len   pad      u8[]     zeros to the bucket size
///
/// The padding is a privacy control. A 1 Hz power stream whose frame length
/// varied with what happened in the house would hand the relay operator the
/// household's load pattern through perfect encryption, so lane 0 frames are
/// always exactly one bucket and never fragment.
public enum Frame {
    public static let version: UInt8 = 1
    public static let headerBytes = 6
    public static let maxPayload = 0xffff

    public static let laneControl: UInt8 = 0
    public static let laneBulk: UInt8 = 1

    /// The delta did not fit its bucket; the rest follows next tick.
    public static let flagTrunc: UInt8 = 0x02

    /// Lane 0 is one fixed size for the life of a session.
    public static let controlBuckets = [256, 512]
    /// Lane 1 steps, because bulk already leaks that a transfer is happening.
    public static let bulkBuckets = [1024, 4096, 16384]

    public struct Decoded: Equatable, Sendable {
        public var lane: UInt8
        public var flags: UInt8
        public var envelope: Envelope

        public var truncated: Bool { flags & Frame.flagTrunc != 0 }
    }

    public struct FrameError: Error, Equatable {
        public let code: String
        public let message: String
    }

    /// Encode one frame, zero-padded to `bucket`.
    ///
    /// Throws rather than growing the bucket: a larger frame than asked for
    /// would defeat the padding entirely.
    public static func encode(lane: UInt8, flags: UInt8 = 0, envelope: Envelope, bucket: Int) throws -> Bytes {
        let payload = encodeCBOR(envelope.cbor)
        if payload.count > maxPayload {
            throw FrameError(code: "E_FRAME_TOO_LARGE", message: "payload \(payload.count) exceeds u16 length field")
        }
        let needed = headerBytes + payload.count
        if needed > bucket {
            throw FrameError(code: "E_FRAME_EXCEEDS_BUCKET", message: "frame needs \(needed) bytes, bucket is \(bucket)")
        }
        var out = Bytes(repeating: 0, count: bucket)
        out[0] = version
        out[1] = lane
        out[2] = flags
        out[3] = 0
        out[4] = UInt8(payload.count >> 8)
        out[5] = UInt8(payload.count & 0xff)
        out.replaceSubrange(headerBytes..<needed, with: payload)
        return out
    }

    /// The smallest bulk bucket the envelope fits, encoded.
    public static func encodeBulk(envelope: Envelope) throws -> Bytes {
        let payload = encodeCBOR(envelope.cbor)
        guard let bucket = bulkBuckets.first(where: { $0 >= payload.count + headerBytes }) else {
            throw FrameError(code: "E_FRAME_EXCEEDS_BUCKET", message: "bulk payload exceeds the largest bucket")
        }
        return try encode(lane: laneBulk, envelope: envelope, bucket: bucket)
    }

    /// Decode one frame. Everything past `len` is ignored without inspection:
    /// the padding is already covered by the AEAD, so checking it would only
    /// add a way to reject a frame that is fine.
    public static func decode(_ bytes: Bytes) throws -> Decoded {
        guard bytes.count >= headerBytes else {
            throw FrameError(code: "E_FRAME_SHORT", message: "frame is \(bytes.count) bytes")
        }
        guard bytes[0] == version else {
            throw FrameError(code: "E_FRAME_VERSION", message: "unsupported frame version \(bytes[0])")
        }
        let len = Int(bytes[4]) << 8 | Int(bytes[5])
        guard headerBytes + len <= bytes.count else {
            throw FrameError(code: "E_FRAME_TRUNCATED", message: "declared length \(len) overruns the frame")
        }
        let value: CBOR
        do {
            value = try decodeCBOR(Array(bytes[headerBytes..<headerBytes + len]))
        } catch {
            throw FrameError(code: "E_FRAME_CBOR", message: "payload is not valid CBOR")
        }
        guard let envelope = Envelope(value) else {
            throw FrameError(code: "E_FRAME_ENVELOPE", message: "envelope must be a map with a string type")
        }
        return Decoded(lane: bytes[1], flags: bytes[2], envelope: envelope)
    }
}

/// One message: a type, an optional request id, an optional body.
public struct Envelope: Equatable, Sendable {
    public var t: String
    public var id: UInt32?
    public var b: CBOR?

    public init(t: String, id: UInt32? = nil, b: CBOR? = nil) {
        self.t = t
        self.id = id
        self.b = b
    }

    /// Unknown keys are ignored: that rule is what lets a newer box talk to
    /// an older app and the reverse.
    init?(_ value: CBOR) {
        guard let t = value["t"]?.string else { return nil }
        self.t = t
        if let raw = value["id"]?.int64, raw >= 0, raw <= Int64(UInt32.max) {
            id = UInt32(raw)
        }
        b = value["b"]
    }

    /// Keys in the web app's order: t, id, b.
    var cbor: CBOR {
        var pairs: [(String, CBOR)] = [("t", .text(t))]
        if let id { pairs.append(("id", .unsigned(UInt64(id)))) }
        if let b { pairs.append(("b", b)) }
        return .map(pairs)
    }
}
